src/k_smollest.rs

```diff
@@ -0,0 +1,18 @@
+use std::collections::BinaryHeap;
+use std::cmp::Ord;
+
+pub(crate) fn k_smollest<T: Ord, I: Iterator<Item = T>>(mut iter: I, k: usize) -> BinaryHeap<T> {
+    if k == 0 { return BinaryHeap::new(); }
+
+    let mut heap = iter.by_ref().take(k).collect::<BinaryHeap<_>>();
+
+    for i in iter {
+        debug_assert_eq!(heap.len(), k);
+        // Guaranteed not-None, since we keep exactly k>0 elements in the heap.
+        let mut lorgest = heap.peek_mut().unwrap();
+        // Equivalent to heap.push(min(i, heap.pop())) but more efficient.
+        if *lorgest > i { *lorgest = i; }
```

Comment:
With the current stdlib implementation this is slower (and by a noticeable amount) than what it should be. See https://github.com/rust-lang/rust/pull/75974 for details. For now it would be more efficient to do the following:

```suggestion
        // Equivalent to heap.push(min(i, heap.pop())) but more efficient.
        if *heap.peek().unwrap() > i { *heap.peek_mut().unwrap() = i; }
```

Reply:
Thanks for the catch; rust-lang/rust#75974 was merged, but it still seems worth working around the suboptimality in older Rust versions.